src/lib.rs

```diff
@@ -411,31 +404,58 @@ cfg_if! {
             );
             Some(mi.assume_init().AllocationBase as usize + get_thread_stack_guarantee() + 0x1000)
         }
-    } else if #[cfg(any(target_os = "linux", target_os="solaris", target_os = "netbsd"))] {
-        unsafe fn guess_os_stack_limit() -> Option<usize> {
-            let mut attr = std::mem::MaybeUninit::<libc::pthread_attr_t>::uninit();
-            assert_eq!(libc::pthread_attr_init(attr.as_mut_ptr()), 0);
-            assert_eq!(libc::pthread_getattr_np(libc::pthread_self(),
-                                                attr.as_mut_ptr()), 0);
-            let mut stackaddr = std::ptr::null_mut();
-            let mut stacksize = 0;
-            assert_eq!(libc::pthread_attr_getstack(
-                attr.as_ptr(), &mut stackaddr, &mut stacksize
-            ), 0);
-            assert_eq!(libc::pthread_attr_destroy(attr.as_mut_ptr()), 0);
-            Some(stackaddr as usize)
+    } else if #[cfg(any(
+        target_os = "linux",
+        target_os = "solaris",
+        target_os = "netbsd",
+        target_os = "freebsd",
+        target_os = "dragonfly",
+        target_os = "illumos"
+    ))] {
+        unsafe fn destroy_pthread_attr(attr: *mut libc::pthread_attr_t) {
+            let ret = libc::pthread_attr_destroy(attr);
+            if ret != 0 {
+                panic!("pthread_attr_destroy failed with error code {}: {}", ret, std::io::Error::last_os_error());
+            }
+        }
+
+        unsafe fn handle_pthread_err(attr: *mut libc::pthread_attr_t, ret: libc::c_int) -> Option<()> {
+            if ret != 0 {
+                destroy_pthread_attr(attr);
+                return None;
+            }
+            Some(())
         }
-    } else if #[cfg(any(target_os = "freebsd", target_os = "dragonfly", target_os = "illumos"))] {
+
         unsafe fn guess_os_stack_limit() -> Option<usize> {
+            // Initialize pthread attributes structure
             let mut attr = std::mem::MaybeUninit::<libc::pthread_attr_t>::uninit();
-            assert_eq!(libc::pthread_attr_init(attr.as_mut_ptr()), 0);
-            assert_eq!(libc::pthread_attr_get_np(libc::pthread_self(), attr.as_mut_ptr()), 0);
+            if libc::pthread_attr_init(attr.as_mut_ptr()) != 0 {
+                return None;
+            }
+
+            let pthread_ctx = attr.as_mut_ptr();
+
+            // Linux and BSD use different functions to get attributes of created thread
+            cfg_if! {
+                if #[cfg(any(target_os = "freebsd", target_os = "dragonfly", target_os = "illumos"))] {
+                    let get_attr = libc::pthread_attr_get_np;
+                } else {
+                    let get_attr = libc::pthread_getattr_np;
+                }
+            };
```

Comment:
Given how compact `guess_os_stack_limit` was before, merging the implementations together only to special case the implementation in the middle makes this part of the change roughly net-negative in my eyes.

`handle_pthread_err` handling resource lifetime(s) is also somewhat iffy -- at this point I'd reach for introducing a wrapper with a destructor for `pthread_attr_t`. Maybe that'll help to clean up the conditional here as well (by e.g. hiding it away into a method?)

If nothing else I would like this to be structured such that the code would not compile if the block of conditions above was modified, but not here, if only to force future maintainer(s) to check this area of code twice.

Reply:
(Stylistically if the lifetime handling of `pthread_attr_t` is moved to be destructor based, then conversions from return code to `Some`/`None` and early returning on error become a fairly straightforward `(res != 0).then_some(())?;`.